File: 92/solution.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        tmp1 = ListNode(0, head)
        tmp2 = tmp1
        for i in range(left - 1):
            tmp2 = tmp2.next
        cur = tmp2.next
        pre = nxt = None
        for _ in range(right - left + 1):
            nxt = cur.next
            cur.next = pre
            pre = cur
            cur = nxt
        tmp2.next = pre
        while tmp2.next:
            tmp2 = tmp2.next
        tmp2.next = nxt
        return tmp1.next

File: 92/Solution.java
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode tmp1 = new ListNode(0, head);
        ListNode tmp2 = tmp1;
        for (int i = 1; i < left; i++) {
            tmp2 = tmp2.next;
        }
        ListNode cur = tmp2.next;
        ListNode pre = null, nxt = null;
        for (int i = 0; i <= right - left; i++) {
            nxt = cur.next;
            cur.next = pre;
            pre = cur;
            cur = nxt;
        }
        tmp2.next = pre;
        while (tmp2.next != null) {
            tmp2 = tmp2.next;
        }
        tmp2.next = nxt;
        return tmp1.next;
    }
}

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}